apps/session-manager/src/utils/lollipop.ts

```diff
@@ -0,0 +1,306 @@
+    O.fromNullable,
+    O.map((pubKeyHashAlgo) => pubKeyHashAlgo.algo),
+    O.getOrElseW(() => void 0 as never),
+  );
+
+const getKeyThumbprintFromSignature = (
+  lollipopSignatureInput: LollipopSignatureInput,
+): E.Either<Errors, Thumbprint> => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [_, thumbprint, ...__] = KEY_ID_REGEX.exec(lollipopSignatureInput) ?? [
+    null,
+    null,
+  ];
+  return Thumbprint.decode(thumbprint);
+};
+
+const getAndValidateAssertionRefForUser = (
+  redisClientSelector: RedisClientSelectorType,
+  fiscalCode: FiscalCode,
+  /* TODO: add event logging
+   * operationId: NonEmptyString,
+   */
+  keyThumbprint: Thumbprint,
+): TE.TaskEither<
+  IResponseErrorInternal | IResponseErrorForbiddenNotAuthorized,
+  AssertionRef
+> =>
+  pipe(
+    RedisSessionStorageService.getLollipopAssertionRefForUser({
+      redisClientSelector,
+      fiscalCode,
+    }),
+    // TODO: send error event if taskEither results to left
+    TE.mapLeft((err) => {
+      log.error(
+        "lollipopMiddleware|error reading the assertionRef from redis [%s]",
+        err.message,
+      );
+      return ResponseErrorInternal("Error retrieving the assertionRef");
+    }),
+    TE.chainW(TE.fromOption(() => ResponseErrorForbiddenNotAuthorized)),
+    TE.chainW(
+      flow(
+        TE.fromPredicate(
+          (assertionRef) =>
+            assertionRef ===
+            `${getAlgoFromAssertionRef(assertionRef)}-${keyThumbprint}`,
+          () => ResponseErrorForbiddenNotAuthorized,
+        ),
+        // TODO: send error event if taskEither results to left
+      ),
+    ),
+  );
+
+/**
+ * Utility function that validate locals to check if all
+ * the properties required for lollipop are present.
+ * The type guard is used to keep unchanged the original locals.
+ * If the type doesn't match a IResponseErrorValidation is returned on left.
+ *
+ * @param locals express res.locals vars injected by toExpressHandler middleware
+ */
+export const withLollipopLocals = <T extends ResLocals>(
+  locals?: T,
+): E.Either<IResponseErrorValidation, LollipopLocalsType> =>
+  pipe(
+    locals,
+    E.fromPredicate(LollipopLocalsType.is, () =>
+      ResponseErrorValidation("Bad request", "Error initializiang lollipop"),
+    ),
+  );
+
+/**
+ * Take a express request and returns in callback validated
+ * lollipop required headers with an `exact` decoding, stripping away
+ * additional headers.
+ * If the validation fails a IResponseErrorValidation is returned.
+ *
+ * @param req the express Request
+ * @param f the callback
+ */
+const withLollipopHeadersFromRequest = async <T>(
+  req: Request,
+  f: (lollipopHeaders: LollipopRequiredHeaders) => Promise<T>,
+): Promise<IResponseErrorValidation | T> =>
+  withValidatedOrValidationError(
+    t.exact(LollipopRequiredHeaders).decode(req.headers),
+    f,
+  );
+
+export const extractLollipopLocalsFromLollipopHeaders = (
```

Comment:
we can remap this function as `RTE` with a single parameters for dependency

Reply:
refactor present in https://github.com/pagopa/io-auth-n-identity-domain/pull/22/commits/0aac1a7af5e32060657c60e53ef458e41e4e445c